### PR TITLE
Improve token creation error message

### LINKED ISSUE #14828: Please improve the error message when creating a token with a project parameter but without a package parameter
**Is your feature request related to a problem? Please describe.**

The following API call leads to a suboptimal error message:
```ShellSession
❯ osc api -X POST "/person/dancermak/token?cmd=create&project=home:dancermak"
Server returned an error: HTTP Error 404: Not Found
Package not found: home:dancermak/
```

**Describe the solution you'd like**

An error message like:
```
The package parameter must be provided in conjunction with the project parameter
```

### PR DESCRIPTION
**Pull request by mfriedrichgh**
Fixes #14828

### CODE REVIEW

File: src/api/app/controllers/person/token_controller.rb [hunk 1]
```diff
@@ -14,7 +14,12 @@ def index
     def create
       authorize @user, :update?
 
-      pkg = (Package.get_by_project_and_name(params[:project], params[:package]) if params[:project] || params[:package])
+      unless params[:project] && params[:package]
+        render_error status: 400, errorcode: 'missing_parameters', message: 'The package and project parameters must be provided.'
```
Comment (krauselukas) [inline src/api/app/controllers/person/token_controller.rb:18]: Hey @mfriedrichgh we have a exception class for this :) so you can use it here https://github.com/openSUSE/open-build-service/blob/ee8f92cfc8134dafe3d541bec26a72588ccd9a91/src/api/lib/api_error.rb#L84

Comment (krauselukas) [inline src/api/app/controllers/person/token_controller.rb:18]:
```suggestion
       raise MissingParameterError, 'The package and project parameters must be provided.'
```

Comment (krauselukas) [reply src/api/app/controllers/person/token_controller.rb:18]: Let me know if something is not clear

Comment (mfriedrichgh) [reply src/api/app/controllers/person/token_controller.rb:18]: Everything clear, thanks for the suggestion.

**Review by krauselukas — CHANGES_REQUESTED** [review]
@mfriedrichgh I just realized that you can create a token without a package or project assigned. Thats why you have a failing test now https://app.circleci.com/pipelines/github/openSUSE/open-build-service/14039/workflows/7444e3f0-8114-483f-8464-40b35203a22c/jobs/145324?invite=true#step-104-182983_103
So basically you would need to check if `project && package` params are given or non of them.

File: src/api/spec/controllers/person/token_controller_spec.rb [hunk 2]
```diff
@@ -53,6 +53,19 @@
...
     end
 
+    context 'called with only project parameter' do
+      before do
+        login user
+      end
+
+      subject { post :create, params: { login: user.login, package: user.home_project, operation: 'rebuild' }, format: :xml }
+
+      it 'creates a global token' do
```
Comment (krauselukas) [inline src/api/spec/controllers/person/token_controller_spec.rb:63]:
```suggestion
      it 'does not create a token' do
```

File: src/api/public/apidocs/paths/person_login_token.yaml [hunk 3]
```diff
@@ -68,19 +68,19 @@ post:
       example: Rebuild gchz
     - in: query
       name: project
+      required: true
       schema:
         type: string
       description: |
-        Project name. Should be provided, together with the package query parameter,
-        to limit the token to a specific package.
+        Project name. Limits the token to a specific project.
```
Comment (krauselukas) [inline src/api/public/apidocs/paths/person_login_token.yaml:75]: Hmm, I would say we should not mark the project and package parameters as required. You can create a global token without providing those parameters. The description was correct before I would say

Comment (mfriedrichgh) [reply src/api/public/apidocs/paths/person_login_token.yaml:75]: Agreed, the edit made before doesn't consider the ability to create token without parameters.

File: src/api/app/controllers/person/token_controller.rb [hunk 4]
```diff
@@ -2,6 +2,7 @@
 module Person
   class TokenController < ApplicationController
     before_action :set_user
+    before_action :set_package, only: [:create]
```
Comment (eduardoj) [inline src/api/app/controllers/person/token_controller.rb:5]: Adding this `before_action` here would call `set_package` even the user is not authorized.

Before the changes of this pull request, checking for the permissions of the user was done before searching for the package. Could we keep this behaviour?

**Review by danidoni — APPROVED** [review]

**Pull request merged**